plover/formatting.py

```diff
@@ -730,44 +794,147 @@ def _atom_to_action_spaces_after(atom, last_action):
+                              last_capitalize, last_upper, last_lower)
+    return action
+
+
+def _apply_mode(text, case, space_char, begin, last_attach,
+                last_glue, last_capitalize, last_upper, last_lower):
+    # Should title case be applied to the beginning of the next string?
+    lower_title_case = ((begin or
+                         last_attach or
+                         last_glue) and not
+                        (last_capitalize or
+                         last_upper))
+    # Apply case, then replace space character
+    text = _apply_case(text, case, lower_title_case)
+    text = _apply_space_char(text, space_char)
+
+    # Title case is sensitive to lower flag
+    if last_lower and len(text) > 0:  # Check for text
+            if (case is _Action.CASES['TITLE']):
+                text = _lower(text)
+
+    return text
+
+
+def _change_mode(command, action):
+
+    """
+    command should be:
+        CAPS, LOWER, TITLE, CAMEL, SNAKE, CLEAR_SPACE,
+            CLEAR_CASE, SET_SPACE or CLEAR
+
+        CAPS: Uppercase
+        LOWER: Lowercase
+        TITLE: Titlecase
+        CAMEL: Titlecase, no space
+        SNAKE: Underscore space
+        CLEAR_SPACE: no space
+        CLEAR_CASE: No case
+        SET_SPACE:xy: Set space to xy
+        CLEAR: No case, no space
+    """
+
+    if command == MODE_CAPS:
+        action.case = _Action.CASES['UPPER']
+
+    elif command == MODE_TITLE:
+        action.case = _Action.CASES['TITLE']
+
+    elif command == MODE_LOWER:
+        action.case = _Action.CASES['LOWER']
+
+    elif command == MODE_SNAKE:
+        action.space_char = '_'
+    elif command == MODE_CAMEL:
+        action.case = _Action.CASES['TITLE']
+        action.space_char = ''
+        action.lower = True
+
+    elif command == MODE_CLEAR:
+        action.space_char = SPACE
+        action.case = _Action.CASES['NONE']
+
+    elif command == MODE_CLEAR_SPACE:
+        action.space_char = SPACE
+
+    elif command == MODE_CLEAR_CASE:
+        action.case = _Action.CASES['NONE']
+
+    elif command.startswith(MODE_SET_SPACE):
+        action.space_char = command[len(MODE_SET_SPACE):]
+
     return action
 
+
+def _apply_case(input_text, case, appended):
+    text = input_text
+    action = _Action()
+    if case is _Action.CASES['LOWER']:
+        text = text.lower()
+    elif case is _Action.CASES['UPPER']:
+        text = text.upper()
+    elif case is _Action.CASES['TITLE']:
+        # Do nothing to appended output
+        if not appended:
+            text = string.capwords(text, " ")
+
+    return text
+
+
+def _apply_space_char(text, space_char):
+    if space_char != ' ':
+        return text.replace(' ', space_char)
+    else:
+        return text
+
+
 def _get_meta(atom):
     """Return the meta command, if any, without surrounding meta markups."""
     if (atom is not None and
         atom.startswith(META_START) and
-        atom.endswith(META_END)):
+            atom.endswith(META_END)):
```

Comment:
Meh, 1 looks ugly, sadly 2 is invalid, 3 is better then... It would have been better if those formatting and whitespace changes had been done in a separate commit: make using `git blame` easier.